Hexlet project 1 DNA (console application)

Svyatoslavik opened the pull request:
Сдача задания

Svyatoslavik committed:
commit my application

Svyatoslavik committed:
add category muttator

Svyatoslavik committed:
small changes

Wenomer commented on HexletProject1DNA/HexletProject1DNA/classes/Cell.m:
```diff
@@ -0,0 +1,64 @@
+    return hamming_distance;    
+}
+
+- (void)mutate:(NSUInteger)persents{
```
Вы немного отступили от задания

```
6. Создайте категорию mutator класса Cell. В ней опишите метод mutate, который возвращает void и принимает int.
```

у вас mutate не в категории

Svyatoslavik replied:
странно я же перед пулом создал категорию, это предыдущая версия, хм

Wenomer commented on HexletProject1DNA/HexletProject1DNA/classes/Cell.m:
```diff
@@ -0,0 +1,64 @@
+        NSInteger idx_of_cells_idx = arc4random() % [indexes_of_cells count];
+        NSInteger replaced_idx = [[indexes_of_cells objectAtIndex:idx_of_cells_idx] intValue];
+        [indexes_of_cells removeObjectAtIndex:idx_of_cells_idx];
+        [_DNA replaceObjectAtIndex:replaced_idx withObject:[self randomDNASymbol]];
```
В этом месте нет гарантии что символ не заменится на точно такойже, в этом случае мутация не произойдет

Svyatoslavik replied:
Спасибо исправил - было бы хорошо если бы такие моменты были освещены в задании потому как сразу не понял что надо заменять именно на другой символ.

Svyatoslavik committed:
Fix mutate method

Fixed problem with replace same sumbols - now with each replaced symbol
will replaced on another symbol (not same)

Svyatoslavik committed:
update project file

GxocT commented:
1. Принято именование переменных и методов вида: имяПеременнойИлиМетода, а не имя_переменной_с_подчеркиванием.
2. Вместо использования NSString в качестве хранилища для нуклеотидов ACTG лучше на мой взгляд использовать NSArray. Во-первых это упростит выборку из массива и код станет более наглядным. Во-вторых, в случае создания DNA из этих нуклеотидов в памяти будут храниться не объекты, а указатели на них, что позволит использовать меньше ОЗУ.
3. Возможно, вместо NSException стоит использовать NSAssert так как он не включается в код при финальной компиляции.

GxocT commented:
Комментарии в Cell.m не помешали бы.

GxocT commented:
Нечитаемо:

```
    NSInteger idx_of_cells_idx = arc4random() % [indexes_of_cells count];
    NSInteger replaced_idx = [[indexes_of_cells objectAtIndex:idx_of_cells_idx] intValue];
    [indexes_of_cells removeObjectAtIndex:idx_of_cells_idx];

    NSMutableString * dna_pattern_string = [NSMutableString stringWithString:DNA_SYMBOLS_ALLOWED];
    [dna_pattern_string deleteCharactersInRange:[dna_pattern_string rangeOfString:[self.DNA objectAtIndex:replaced_idx]]];        
    [self.DNA replaceObjectAtIndex:replaced_idx withObject: [dna_pattern_string substringWithRange:NSMakeRange(arc4random() % [dna_pattern_string length], 1)] ];
```

Либо разбейте на более мелкие элементы, либо откомментируйте.
